### DISCUSSION TITLE
How to use pillow_heif in Dataloader (Pytorch)

### DISCUSSION THREAD

**Discussion by qilong-zhang** [bigcat88/pillow_heif#70] [Q&A] [answered]
Hi @bigcat88 , I want to load ".heic" images in Dataloader. So my code is 
```python
Class Dataset(torch.utils.data.Dataset):
    def __init__(self, path=None, mode='train'):
          pass
    def __len__(self):
          pass
    def __getitem__(self, index):
          register_heif_opener()
          pass
```
But this code seems to cause a memory leak. Is there any other better solution?

**Comment by bigcat88** [comment]
Can you provide fuller example?

1. `register_heif_opener` can be called only once somewhere when program starts
2. very strange that `register_heif_opener` causes a memory leak. What version of Pillow are you using?

**Reply by qilong-zhang** [reply]
> Can you provide fuller example?
> 
> 1. `register_heif_opener` can be called only once somewhere when program starts
> 2. very strange that `register_heif_opener` causes a memory leak. What version of Pillow are you using?
```python
from multiprocessing import shared_memory
import glob
from pillow_heif import register_heif_opener
import torch

class Dataset_Test(torch.utils.data.Dataset):                                    
    def __init__(self, path=None):
        self.transform = ToTensor()
        self.common_path = glob.glob(os.path.join(path, '*', '*'))
        self.common_path = shared_memory.ShareableList(self.common_path)

    def __len__(self):
        return len(self.common_path)

    def __getitem__(self, index):        
        register_heif_opener()                              
        name = self.common_path[index]
        image = Image.open(name).convert('RGB')
        
        return self.transform(image)
```
Here is my complete code, the description just given about the memory leak may not be accurate. My main problem at the moment is that I find that if I call register_heif_opener() every time inside "\__getitem__",  it will load slower, is there any way to call it only once?

Besides, If I use the above code, after several iterations, the running code will terminate, and report:
```python
Traceback (most recent call last):
  File "train.py", line 289, in <module>
    run_demo(main, n_gpus)
  File "train.py", line 204, in run_demo
    mp.spawn(demo_fn,
  File "/root/mambaforge3/envs/torch-gpu/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 240, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
  File "/root/mambaforge3/envs/torch-gpu/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 198, in start_processes
    while not context.join():
  File "/root/mambaforge3/envs/torch-gpu/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 140, in join
    raise ProcessExitedException(
torch.multiprocessing.spawn.ProcessExitedException: process 2 terminated with signal SIGKILL
```

**Comment by bigcat88** [comment, marked as answer]
You can call it before the definition of `Dataset_Test` class for example.

```python3
from multiprocessing import shared_memory
import glob
from pillow_heif import register_heif_opener
import torch

register_heif_opener() 

class Dataset_Test(torch.utils.data.Dataset):  
```

if it works, you can simplify it even more with one import line and without calling `register_heif_opener`:
```python3
from multiprocessing import shared_memory
import glob
import pillow_heif.HeifImagePlugin  # noqa
import torch
```

**Reply by qilong-zhang** [reply]
This example can run, but it still report the same error and the loading speed is still slow.

**Reply by bigcat88** [reply]
You can try `register_heif_opener(decode_threads=8)`, by default `libheif` uses only 4 threads for image decoding (but not all images support decoding in multiply threads, e.g. not all images have more than one tile).

**Reply by qilong-zhang** [reply]
Thank you @bigcat88. `register_heif_opener(decode_threads=8)` does help. But if I want to avoid the `SIGKILL`, it seems that I need to increase the memory: If memory=30G, my code will termate in 100 itearations. But if I set memory=100G, my code is still running at the moment and no errors are reported.

**Reply by bigcat88** [reply]
Maybe it just threw an exception while decoding some image (which results in `SIGKILL`), in which case I need that image for research. 
But that's just guesswork.

**Reply by qilong-zhang** [reply]
One of the findings is that the memory grows slightly as the number of iterative steps increases.

**Reply by bigcat88** [reply]
in my local tests i do something like million of iterations on opening and decoding images and did not found any leaks for last 2 versions.
but if you can provide some minimal example to investigate, i will be glad to fix a problem if it is present.

P.S.: and yeh, when calling `register_heif_opener` in cycle their will be leak in Pillow, but probably it is just invalid usage, I will ask in their repo about this later today.

**Reply by qilong-zhang** [reply]
Thanks! My code aims to train a classfication model with pytorch DDP (4 GPUs). Therefore, it is not easy to provide a minimal example, as this also requires large-scale data sets. But one thing is for sure, if I don't consider heic format images, I won't have these problems.

Interestingly, it seems that the memory will no longer continue to rise slightly when the model continues to be trained after finishing one epoch training.

**Reply by qilong-zhang** [reply]
> Thanks! My code aims to train a classfication model with pytorch DDP (4 GPUs). Therefore, it is not easy to provide a minimal example, as this also requires large-scale data sets.
> 
> Interestingly, it seems that the memory will no longer continue to rise slightly when the model continues to be trained after finishing one epoch training.

Not right, the memory still rise silightly..

**Reply by bigcat88** [reply]
Without an example in format:
```python3
for i in range(1000):
    image = Image.open("path_to_image").convert('RGB')
    image = self.transform(image)
```
i probably can not help.
Are you sure that this problem is not present without `heic` images?(saw that you mention that there is no problem when you do not open HEIF images)

P.S. _(not related to current problem_): in next Pillow(9.5.0) they will add support for multiply calls of `register_heif_opener` 
python-pillow/Pillow#6916

**Reply by qilong-zhang** [reply]
I retest my code and I find that if I don't train the model, the memory doesn't grow; but if I train the model, the memory grows regardless of whether it's 1 GPU or multiple GPUs. Fortunately, memory grows relatively slowly, so for now I don't have `SIGKILL` as long as I have enough memory.

**Reply by qilong-zhang** [reply]
> Without an example in format:
> 
> ```python
> for i in range(1000):
>     image = Image.open("path_to_image").convert('RGB')
>     image = self.transform(image)
> ```
> 
> i probably can not help. Are you sure that this problem is not present without `heic` images?(saw that you mention that there is no problem when you do not open HEIF images)
> 
> P.S. _(not related to current problem_): in next Pillow(9.5.0) they will add support for multiply calls of `register_heif_opener` [python-pillow/Pillow#6916](https://github.com/python-pillow/Pillow/issues/6916)

I checked my code again. I found that the memory growth was a problem that was already present in my code. If I consider `.heic` format images, the memory grows relatively faster, thus causing `SIGKILL`. So for now, it looks like the problem may not be fundamentally caused by `.heic`.
